server/src/main/java/org/opensearch/index/mapper/IpFieldMapper.java

```diff
@@ -225,42 +226,49 @@ protected Object parseSourceValue(Object value) {
             failIfNotIndexedAndNoDocValues();
-            Query query;
+            final PointRangeQuery pointQuery;
             if (value instanceof InetAddress) {
-                query = InetAddressPoint.newExactQuery(name(), (InetAddress) value);
+                pointQuery = (PointRangeQuery) InetAddressPoint.newExactQuery(name(), (InetAddress) value);
             } else {
                 if (value instanceof BytesRef) {
                     value = ((BytesRef) value).utf8ToString();
                 }
                 String term = value.toString();
                 if (term.contains("/")) {
                     final Tuple<InetAddress, Integer> cidr = InetAddresses.parseCidr(term);
-                    query = InetAddressPoint.newPrefixQuery(name(), cidr.v1(), cidr.v2());
+                    pointQuery = (PointRangeQuery) InetAddressPoint.newPrefixQuery(name(), cidr.v1(), cidr.v2());
                 } else {
                     InetAddress address = InetAddresses.forString(term);
-                    query = InetAddressPoint.newExactQuery(name(), address);
+                    pointQuery = (PointRangeQuery) InetAddressPoint.newExactQuery(name(), address);
                 }
             }
-            if (isSearchable() && hasDocValues()) {
-                String term = value.toString();
-                if (term.contains("/")) {
-                    final Tuple<InetAddress, Integer> cidr = InetAddresses.parseCidr(term);
-                    return InetAddressPoint.newPrefixQuery(name(), cidr.v1(), cidr.v2());
-                }
-                return new IndexOrDocValuesQuery(
-                    query,
-                    SortedSetDocValuesField.newSlowExactQuery(name(), new BytesRef(((PointRangeQuery) query).getLowerPoint()))
+            Query dvQuery = null;
+            if (hasDocValues()) {
+                dvQuery = SortedSetDocValuesField.newSlowRangeQuery(
+                    name(),
+                    new BytesRef(pointQuery.getLowerPoint()),
+                    new BytesRef(pointQuery.getUpperPoint()),
+                    true,
+                    true
                 );
             }
-            if (hasDocValues()) {
-                String term = value.toString();
-                if (term.contains("/")) {
-                    final Tuple<InetAddress, Integer> cidr = InetAddresses.parseCidr(term);
-                    return InetAddressPoint.newPrefixQuery(name(), cidr.v1(), cidr.v2());
+            return indexOrDvQuery(pointQuery, dvQuery);
+        }
+
+        private Query indexOrDvQuery(Query pointQuery, Query dvQuery) {
+            if (isSearchable() && hasDocValues()) {
+                assert pointQuery != null;
+                assert dvQuery != null;
+                return new IndexOrDocValuesQuery(pointQuery, dvQuery);
+            } else {
+                if (isSearchable()) {
+                    assert pointQuery != null;
+                    return pointQuery;
+                } else {
```

Comment:
```suggestion
                } else if (hasDocValues()) {
```?

Reply:
I see. But adding a condition here will require to return something in alt branch that seems verbose to me. I suppose it's ok for this private method to rely on enclosing methods invariant `failIfNotIndexedAndNoDocValues();`